### PR TITLE
Refactor Notifications to decouple business logic and HTTP server

### PR DESCRIPTION
**Pull request by nkstanchev**
…logic

Addresses #1579.

Signed-off-by: Nikolay Stanchev <nstanchev@vmware.com>

### CODE REVIEW

**Comment by brandonforster** [comment]
related to #941

File: internal/support/notifications/rest_notification.go [hunk 1]
```diff
@@ -218,17 +341,11 @@ func notificationBySenderHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	n, err := dbClient.GetNotificationBySender(vars["sender"], limitNum)
+	n, err := getNotificationsBySender(vars["sender"], limitNum)
 	if err != nil {
-		if err == db.ErrNotFound {
```
Comment (brandonforster) [inline internal/support/notifications/rest_notification.go]: Why did you remove this `ErrNotFound` check?

File: internal/support/notifications/rest_notification.go [hunk 2]
```diff
@@ -265,16 +382,12 @@ func notificationByStartEndHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	n, err := dbClient.GetNotificationsByStartEnd(start, end, limitNum)
+	n, err := getNotificationsByStartEnd(start, end, limitNum)
 	if err != nil {
-		if err == db.ErrNotFound {
```
Comment (brandonforster) [inline internal/support/notifications/rest_notification.go]: Why did you remove this `ErrNotFound` check?

File: internal/support/notifications/rest_subscription.go [hunk 3]
```diff
@@ -22,29 +22,156 @@ import (
...
 	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
+	"github.com/edgexfoundry/edgex-go/internal/support/notifications/errors"
 	"github.com/edgexfoundry/go-mod-core-contracts/models"
 	"github.com/gorilla/mux"
 )
 
-const (
-	applicationJson = "application/json; charset=utf-8"
-)
+func getSubscriptions() (subscriptions []models.Subscription, err error) {
```
Comment (brandonforster) [inline internal/support/notifications/rest_subscription.go]: One of the goals of this effort was to decouple this completely. Could you move these functions into their own file, so that the rest logic and the business logic don't live together?

File: internal/support/notifications/rest_notification.go [hunk 4]
```diff
@@ -17,16 +17,169 @@ package notifications
...
 	"strconv"
 
 	"github.com/edgexfoundry/edgex-go/internal/pkg"
 	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
+	"github.com/edgexfoundry/edgex-go/internal/support/notifications/errors"
 	"github.com/edgexfoundry/go-mod-core-contracts/models"
 	"github.com/gorilla/mux"
 )
 
+func addNotification(n *models.Notification) (e error) {
```
Comment (brandonforster) [inline internal/support/notifications/rest_notification.go]: Please move these functions to their own file.

File: internal/support/notifications/rest_notification.go [hunk 5]
```diff
@@ -17,16 +17,169 @@ package notifications
...
+	n, err = dbClient.GetNewNotifications(limit)
+	if err != nil {
+		LoggingClient.Error(err.Error())
+		return n, err
+	}
+	return n, nil
+}
+
 func notificationHandler(w http.ResponseWriter, r *http.Request) {
+
```
Comment (brandonforster) [inline internal/support/notifications/rest_notification.go]: Please remove this newline.

File: internal/support/notifications/rest_notification.go [hunk 6]
```diff
@@ -41,33 +194,25 @@ func notificationHandler(w http.ResponseWriter, r *http.Request) {
...
+	}
+	err = checkSeverity(&n)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
 	}
 
 	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
 	w.WriteHeader(http.StatusAccepted)
+	fmt.Println(n.ID)
```
Comment (brandonforster) [inline internal/support/notifications/rest_notification.go]: Please remove this.

File: internal/support/notifications/rest_notification.go [hunk 7]
```diff
@@ -385,16 +484,9 @@ func notificationsByLabelsHandler(w http.ResponseWriter, r *http.Request) {
...
-	n, err := dbClient.GetNotificationsByLabels(labels, limitNum)
+	n, err := getNotificationsByLabels(labels, limitNum)
 	if err != nil {
-		if err == db.ErrNotFound {
-			http.Error(w, "Notification not found", http.StatusNotFound)
-		} else {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-		}
-		LoggingClient.Error(err.Error())
-		w.Header().Set("Content-Type", applicationJson)
```
Comment (brandonforster) [inline internal/support/notifications/rest_notification.go]: Please do not remove the header set.

File: internal/support/notifications/rest_subscription.go [hunk 8]
```diff
@@ -117,40 +250,33 @@ func subscriptionByIDHandler(w http.ResponseWriter, r *http.Request) {
...
-		LoggingClient.Info("Deleting subscription: " + id)
-
-		if err = dbClient.DeleteSubscriptionById(id); err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			LoggingClient.Error(err.Error())
 			return
 		}
 		w.Header().Set("Content-Type", "application/json")
 		w.WriteHeader(http.StatusOK)
+		w.Write([]byte("true"))
```
Comment (brandonforster) [inline internal/support/notifications/rest_subscription.go]: Why did you add this?

**Comment by tsconn23** [comment]
I have not reviewed this PR and I'm not necessarily opposed to it. But I will say it would be of greater benefit to the project to adopt the Operator pattern being used in core-metadata to facilitate greater test coverage. There are no tests in this PR so it doesn't move us any closer to one of our Fuji deliverable goals.

Also please fix the commit message
https://chris.beams.io/posts/git-commit/

File: internal/support/notifications/rest_subscription.go (hunk 8 above)
Comment (nkstanchev) [reply internal/support/notifications/rest_subscription.go]: Because I saw it was like this in core data so I added it for consistency

File: internal/support/notifications/rest_notification.go (hunk 7 above)
Comment (nkstanchev) [reply internal/support/notifications/rest_notification.go]: I think it is not necessary in this case. http.Error handles the Content-Type

File: internal/support/notifications/rest_notification.go (hunk 2 above)
Comment (nkstanchev) [reply internal/support/notifications/rest_notification.go]: This function does not return a ErrNotFound, it returns an empty json array when it doesn't find anything

File: internal/support/notifications/rest_notification.go (hunk 1 above)
Comment (nkstanchev) [reply internal/support/notifications/rest_notification.go]: This function does not return a ErrNotFound, it returns an empty json array when it doesn't find anything

File: internal/support/notifications/rest_notification.go (hunk 6 above)
Comment (nkstanchev) [reply internal/support/notifications/rest_notification.go]: Sure

**Comment by nkstanchev** [comment]
I was not aware that we are going for the Operator pattern for this service. In this case, I'll start working on this transition.
